pkg/service/service.go

```diff
@@ -1072,7 +1077,27 @@ func (s *Service) restoreAndDeleteOrphanSourceRanges() error {
 //
 // The removal is based on an assumption that during the sync period
 // UpsertService() is going to be called for each alive service.
-func (s *Service) SyncWithK8sFinished() error {
+func (s *Service) SyncWithK8sFinished(ensurer func(k8s.ServiceID, *lock.StoppableWaitGroup) bool) error {
```

Comment:
Did you consider making this function return the set of services and then call `ensurer` on the return of `SyncWithK8sFinished` instead? Might make it easier to understand what `ensurer` is for. Anonymous functions can make the code a bit more convenient to write, but hinders readability. This is a minor comment though because there's only one call site of `SyncWithK8sFinished`, so mostly just wondering what you've considered.

Reply:
I personally don't have a strong opinion here. IMO the current approach has the advantage that the entire synchronization logic is self-contained in a single function, which is then easier to be tested. Another possibility might be to propagate the entire service cache (rather than the single function) as parameter (but that would make testing harder), or replace the anonymous function with an interface (I initially discarded this option as being more verbose, but it might improve clarity). WDYT?

Reply:
Given that from an offline conversation about this function we talked about this function is executed in a controller although it's one-off, maybe we just defer the refactoring of it anyway (into a hive job or something). I'm fine as it is.